kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -305,22 +316,37 @@ protected void handleTopicMetadataRequest(KafkaHeaderAndRequest metadataHar,
         //      e.g. <topic1, {persistent://public/default/topic1-partition-0,...}>
         //   1. no topics provided, get all topics from namespace;
         //   2. topics provided, get provided topics.
-        CompletableFuture<Map<String, List<TopicName>>> pulsarTopicsFuture =
-            (topics == null || topics.isEmpty())
-                ?
-                pulsarService.getNamespaceService().getListOfPersistentTopics(namespace)
-                    .thenApply(
-                        list -> list.stream()
-                            .map(topicString -> TopicName.get(topicString))
-                            .collect(Collectors
-                                .groupingBy(
-                                    topicName -> getKafkaTopicNameFromPulsarTopicname(topicName),
-                                    Collectors.toList()))
-                    )
-                :
-                new CompletableFuture<>();
-
-        if (!(topics == null || topics.isEmpty())) {
+        CompletableFuture<Map<String, List<TopicName>>> pulsarTopicsFuture = new CompletableFuture<>();
+
+        if (topics == null || topics.isEmpty()) {
+            // get all topics
+            Map<String, List<TopicName>> pulsarTopics = Maps.newConcurrentMap();
+            final String offsetsTopicName = new KopTopic(String.join("/",
+                    kafkaConfig.getKafkaMetadataTenant(),
+                    kafkaConfig.getKafkaMetadataNamespace(),
+                    Topic.GROUP_METADATA_TOPIC_NAME)
+            ).getFullName();
+            try {
+                for (String tenant : pulsarService.getAdminClient().tenants().getTenants()) {
+                    for (String namespace : admin.namespaces().getNamespaces(tenant)) {
```

Comment:
since we are inside broker, `pulsarService.getNamespaceService()` would be a better choice than admin(which is a client); if not, please use async operations to avoid thread waiting and hang,

Reply:
It seems that `NamespaceService` can't access all tenants/namespaces. I'll implement the async version soon.

Reply:
If not Ok, we could use admin.ns().xxxAsync()?

Reply:
Yeah, I just mean that.

Reply:
It uses async operations now, and a unit test was added to `DifferentNamespaceTest`.